src/operators/reduce.rs

```diff
@@ -448,10 +450,10 @@ where
                         // TODO: It would be better if all updates went into one batch, but timely dataflow prevents
                         //       this as long as it requires that there is only one capability for each message.
                         let mut buffers = Vec::<(G::Timestamp, Vec<(V, G::Timestamp, T2::Diff)>)>::new();
-                        let mut builders = Vec::new();
+                        let mut chains = Vec::new();
```

Comment:
Worth noting that this is a potential regression, on account of staging all the data. You mentioned that another pattern is just to flush the buffers .. whenever one likes. I figure we should either jot that down in a comment, or make it be the case if we are worried about it.

Reply:
I looked into whether it's possible to populate a builder from `((&K, V), T, R)` instead of cloning `K`. The short answer is no, the long answer is maybe. At the moment, we have two blockers:
1. Reductions write their output into an arrangement, but bypass the merge batcher. We still need to provide a merge batcher type, even tho it'll never be used. We can test removing it by providing a dummy `Batcher` trait implementation that is mostly `unimplemented!()`.
2. The input chunks to the builder need to implement `Container`, which implies `'static`. No way to store references! https://github.com/TimelyDataflow/timely-dataflow/pull/540 might help here.

I think this is something we should keep as a to-do, but not prioritize until we have evidence it causes performance issues.